HealthVault/Classes/Caching/MHVThingCacheDatabase.m

```
@@ -0,0 +1,448 @@
+    __block NSString *recordId;
+    
+    // Make sure properties are retrieved with the correct context
+    [self.managedObjectContext performBlockAndWait:^
+     {
+         recordId = ((MHVCachedRecord *)record).recordId;
+     }];
+    return recordId;
+}
+
+- (NSDate *)lastSyncDateFromRecord:(id<MHVCachedRecord>)record
+{
+    __block NSDate *date;
+    
+    // Make sure properties are retrieved with the correct context
+    [self.managedObjectContext performBlockAndWait:^
+     {
+         date = ((MHVCachedRecord *)record).lastSyncDate;
+     }];
+    return date;
+}
+
+- (NSInteger)lastSequenceNumberFromRecord:(id<MHVCachedRecord>)record
+{
+    __block NSInteger sequenceNumber;
+    
+    // Make sure properties are retrieved with the correct context
+    [self.managedObjectContext performBlockAndWait:^
+     {
+         sequenceNumber = ((MHVCachedRecord *)record).lastOperationSequenceNumber;
+     }];
+    return sequenceNumber;
+}
+
+- (void)updateRecord:(id<MHVCachedRecord>)record lastSyncDate:(NSDate *)lastSyncDate sequenceNumber:(NSNumber *)sequenceNumber
+{
+    // Make sure properties are updated with the correct context
+    [self.managedObjectContext performBlockAndWait:^
+     {
+         MHVCachedRecord *cachedRecord = (MHVCachedRecord *)record;
+         
+         if (lastSyncDate)
+         {
+             cachedRecord.lastSyncDate = lastSyncDate;
+         }
+         if (sequenceNumber)
+         {
+             cachedRecord.lastOperationSequenceNumber = sequenceNumber.integerValue;
+         }
+         
+         [self saveContext];
+     }];
+}
+
+#pragma mark - Core Data
+
+- (NSError *)saveContext
+{
+    if (self.deleted)
+    {
+        return nil;
+    }
+    
+    __block NSError *error = nil;
+    
+    [self.managedObjectContext performBlockAndWait:^
+    {
+        if ([self.managedObjectContext hasChanges] && ![self.managedObjectContext save:&error])
+        {
+            MHVLOG(@"Error: %@", error.localizedDescription);
```

It isn't sufficient for us to just log the error, we need to return this error to the caller for debugging purposes and testing purposes.